File: pa1/Lex.java
/*
 * Lex.java
 *
 * Copyright (C) 2017, Spencer Peterson
 */

import java.io.*;
import java.util.Scanner;
import java.util.ArrayList;

public class Lex {
	public static void main(String args[]) throws IOException {
		Scanner input = null;
		PrintWriter output = null;
		List list = new List();
		ArrayList<String> lines = new ArrayList<String>();

		// Check for useage
		if (args.length != 2) {
			System.out.println("Usage: infile outfile");
			System.exit(-1);
		}

		// Setup input/output
		input = new Scanner(new File(args[0]));
		output = new PrintWriter(new FileWriter(args[1]));

		// Read in the input
		while (input.hasNextLine()) {
			lines.add(input.nextLine());
		}

		// Iterate through the lines
		for (int i = 0; i < lines.size(); i++) {

			// Start each iteration at the front
			list.moveFront();

			// Scrub the list to where the line belongs
			while (list.index() != -1 
					&& lines.get(i).compareTo(lines.get(list.get())) >= 0) {
				list.moveNext();
			}

			// If the index went off, then append the word
			if (list.index() == -1) {
				list.append(i);
			}

			// Otherwise insert it before the word we're at
			else {
				list.insertBefore(i);
			}
		}

		// Print it out
		for (int i : list) {
			output.println(lines.get(i));
		}

		// Clean up
		input.close();
		output.close();
	}
}

File: pa2/Lex.c
/*
 * Lex.c
 *
 * This sorts a file's lines alphabetically using a list.
 *
 * Author: Spencer Peterson
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "List.h"

#define LINE_BUFFER_SIZE 256

char **readLines(FILE *input);

int main(int argc, char **argv) {
    FILE *in = NULL, *out = NULL;    
    List list = newList();
    char **lines = NULL;
    int i;

    // Check usage
    if (argc != 3) {
        printf("Usage: %s infile outfile\n", argv[0]);
        exit(1);
    }

    // Open files
    in = fopen(argv[1], "r");
    out = fopen(argv[2], "w");

    // Make sure files opened
    if (in == NULL || out == NULL) {
        printf("Could not open file(s)");
        exit(1);
    }

    // Read in the input
    lines = readLines(in);

    // Sort the lines into the list
    for (i = 0; lines[i] != NULL; i++) {

        moveFront(list);

        // Scrub to location the line belongs
        while (index(list) != -1 && strcmp(lines[i], lines[get(list)]) >= 0) {
            moveNext(list);
        }

        // If index is -1, then this line belongs at the end
        if (index(list) == -1) {
            append(list, i);
        }
        // Otherwise put it right before where we stopped
        else {
            insertBefore(list, i);
        }
    }

    // Print out the sorted list into the output file
    moveFront(list);
    while (index(list) != -1) {
        fprintf(out, "%s", lines[get(list)]);
        moveNext(list);
    }

    // Free the lines
    for (i = 0; lines[i] != NULL; i++) {
        free(lines[i]);
    }
    free(lines);

    // Free the list
    freeList(&list);

    // Close the files
    fclose(in);
    fclose(out);
}

char **readLines(FILE *input) {
    char **lines = NULL;
    char line[LINE_BUFFER_SIZE];
    int lineslen = 0;

    // Read lines in a loop
    while (fgets(line, LINE_BUFFER_SIZE, input) != NULL) {
        // Lengthen the lines array
        lineslen++;
        lines = realloc(lines, lineslen*sizeof(char*));

        // Allocate memory for the line
        lines[lineslen-1] = malloc(sizeof(char)*(strlen(line)+1));

        // Copy it in from the buffer
        strcpy(lines[lineslen-1], line);
    }

    // Add a terminator
    lineslen++;
    lines = realloc(lines, lineslen*sizeof(char*));
    lines[lineslen-1] = NULL;

    return lines;
}
